### DISCUSSION TITLE
How to generate mesh from point coordinates or Points

### DISCUSSION THREAD

**Discussion by DRLing2021** [marcomusy/vedo#784] [Q&A] [answered]
Hello:

We have a set of 3D point coordinates to plot and analyze. We've searched the vedo API and community and found that using Points([X,Y,Z]) can easily convert coordinates to Points object in Vedo. However, there are some useful methods are based on Mesh objects. We  expect a method to convert Points to Mesh but see no  examples. Can anyone give us a hand? 

Your advice is deeply appreciated.
Regards,
DR, Ling

**Comment by ZiguoAtGitHub** [comment]
Hi Ling
I guess you can try delaunay3d function to create 3D Delaunay triangulation of input points and then get the Mesh.
Ziguo

**Comment by marcomusy** [comment, marked as answer]
In principle you can also create a `Mesh` object from 3D coordinates with eg:
```python
from vedo import *
pts = np.random.randn(100,3)
msh = Mesh(Points(pts), c='black').point_size(10)
show(msh, axes=1)
```
what are the method which you need from the Mesh class? In theory there should be none..
Hope this helps!

**Comment by DRLing2021** [comment]
First thank you Marcomusy and Ziguo!

The mesh method I was intended to use is  compute_curvature(). Did I misunderstand the online document? 

BTW, what does the point_size() mean? It is hard for me to search the online document to find what I am looking for. Are there any offline pdf format document available? 

Regards!

**Reply by DRLing2021** [reply]
I've organized the following code to test. Unfortunately the  face_list is empty that makes me think the mesh is not generated. Based on https://vedo.embl.es/autodocs/content/vedo/mesh.html, the methods , compute_curvature() and boundaries(), need a mesh created first. So how to create a mesh from pts coords?

we also searched on https://vedo.embl.es/autodocs/content/vedo/pointcloud.html but find no informaton about the delaunay3d  function but a delaunay2d function which seems not applicable  

Did I misunderstand something in the documentation? Your help is appreciated.
Regards.

```python
def main():
    print("OK")
    Gen_PC()

def Gen_PC():
    X = np.arange(-10, 10, 1)
    Y = np.arange(-10, 10, 1)
    X, Y = np.meshgrid(X, Y)
    R = np.sqrt(X ** 2 + Y ** 2)
    Z = np.sin(R)

    X_flat = np.reshape(X,-1)
    Y_flat = np.reshape(Y,-1)
    Z_flat = np.reshape(Z,-1)

    xyz = np.zeros((np.size(X), 3))
    xyz[:, 0] = X_flat
    xyz[:, 1] = Y_flat
    xyz[:, 2] = Z_flat

    print("data len is ",len(X_flat))
    show_PC = Points([X_flat,Y_flat,Z_flat],c="green")
    show_mesh = Mesh(show_PC, c='black').point_size(10)

    print("mesh type is ",type(show_mesh))
    face_list = show_mesh.faces()
    print("face list is ",face_list)
```

**Reply by marcomusy** [reply]
Hi, you can only compute the curvature if have the information of the connectivity of the point cloud - otherwise it is not mathematically unambiguously defined.
Hence you can try to reconstruct the surface Mesh and from there you  will be able to compute curvatures, (see 
 `examples/advanced/recosurface.py`).

Or you can try to guess the local structure of your point cloud by computing point normals and from there guess the local curvatures. 
This could be a starting point:
```python

from vedo import *

s = Sphere(quads=True, res=10) # some test points in space
pts = s.points()

vpts = Points(pts)
vpts.compute_normals_with_pca(invert=True)
vpts.print()

normals = vpts.pointdata["Normals"]
arrows  = Arrows(pts, pts + normals/10).c('red5')

show(vpts, arrows, axes=True).close()
```
![Screenshot from 2023-01-09 14-10-41](https://user-images.githubusercontent.com/32848391/211316624-d78710a0-7571-409f-a740-58dbd271583c.png)

**Reply by DRLing2021** [reply]
Thank you Marco.  

With you suggestion I read through the recosurface sample and the API reference online. I have questions about dim:
1. if only one dim set, this value applies to all 3 dimensions isn't it?
2.  What is the exactly meaning of dim:  Is it the resolution count at given dimension? eg if x value range is [0,200], dim = 100, then the x resultion is  200/100 = 2. 

At the same time, I am thinking about using Delaunay function of scipy to generate the triangles and use the simplices as the face list of Vedo Mesh input. Is that a valid path to do it too? 

Regards.

**Reply by marcomusy** [reply]
1. Yes
2. yes, it is the number of voxels in x, y and z to control precision
 
You can use delaunay in vedo too check out:
`examples/basic/delaunay2d.py`
you can choose to use `mode='fit'`.

If your point cloud is convex you can also use `ConvexHull()`  as in
`examples/advanced/convexHull.py`

Let me know if you need more help.
